File: backend/app/models/asset_history.py
"""AssetHistory ORM model."""

import uuid
from datetime import datetime, timezone

from sqlalchemy import DateTime, ForeignKey, String, Text
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import Mapped, mapped_column

from app.database.base import Base


class AssetAction:
    CREATED = "created"
    UPDATED = "updated"
    ASSIGNED = "assigned"
    RETURNED = "returned"
    TRANSFERRED = "transferred"
    STATUS_CHANGED = "status_changed"
    CONDITION_CHANGED = "condition_changed"


class AssetHistory(Base):
    __tablename__ = "asset_histories"

    id: Mapped[str] = mapped_column(
        UUID(as_uuid=False), primary_key=True, default=lambda: str(uuid.uuid4())
    )
    asset_id: Mapped[str] = mapped_column(
        UUID(as_uuid=False),
        ForeignKey("assets.id", ondelete="RESTRICT"),
        nullable=False,
        index=True,
    )
    action: Mapped[str] = mapped_column(
        String(30), nullable=False, index=True
    )
    performed_by: Mapped[str | None] = mapped_column(
        UUID(as_uuid=False),
        ForeignKey("users.id", ondelete="SET NULL"),
        nullable=True,
        index=True,
    )
    performed_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=False),
        default=lambda: datetime.now(timezone.utc).replace(tzinfo=None),
        index=True,
    )
    old_values: Mapped[str | None] = mapped_column(Text, nullable=True)
    new_values: Mapped[str | None] = mapped_column(Text, nullable=True)
    notes: Mapped[str | None] = mapped_column(Text, nullable=True)

File: assetoptima_field/lib/features/assets/models/asset_history.dart
class AssetHistory {
  final String id;
  final String action;
  final String? performedBy;
  final DateTime performedAt;
  final String? notes;

  const AssetHistory({
    required this.id,
    required this.action,
    this.performedBy,
    required this.performedAt,
    this.notes,
  });

  factory AssetHistory.fromJson(Map<String, dynamic> json) => AssetHistory(
        id: json['id'] as String,
        action: json['action'] as String,
        performedBy: json['performed_by'] as String?,
        performedAt: DateTime.parse(json['performed_at'] as String),
        notes: json['notes'] as String?,
      );

  String get actionLabel {
    switch (action) {
      case 'create':
        return 'Created';
      case 'update':
        return 'Updated';
      case 'assign':
        return 'Assigned';
      case 'return':
        return 'Returned';
      case 'transfer':
        return 'Transferred';
      case 'maintenance':
        return 'Maintenance';
      case 'verify':
        return 'Verified';
      default:
        return action;
    }
  }
}
